
<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo">
</center>


# Access DB2 on Cloud using Python


Estimated time needed: **15** minutes
    

## Objectives

After completing this lab you will be able to:

* Create a table
* Insert data into the table
* Query data from the table
* Retrieve the result set into a pandas dataframe
* Close the database connection



# Accéder à DB2 sur le Cloud à l’aide de Python

Temps estimé nécessaire : **15** minutes
    
## Objectifs

À l’issue de cet atelier, vous serez en mesure de :

* Créer un tableau
* Insérer les données dans le tableau
* Interroger les données de la table
* Récupérer le jeu de résultats dans un dataframe pandas
* Fermez la connexion à la base de données




__Notice:__ Please follow the instructions given in the first Lab of this course to Create a database service instance of Db2 on Cloud.

## Task 1: Import the `ibm_db` Python library

The `ibm_db` [API ](https://pypi.python.org/pypi/ibm_db/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork22-2022-01-01) provides a variety of useful Python functions for accessing and manipulating data in an IBM® data server database, including functions for connecting to a database, preparing and issuing SQL statements, fetching rows from result sets, calling stored procedures, committing and rolling back transactions, handling errors, and retrieving metadata.


We import the ibm_db library into our Python Application

The following required modules are pre-installed in the Skills Network Labs environment. However if you run this notebook commands in a different Jupyter environment (e.g. Watson Studio or Ananconda) you may need to install these libraries by removing the `#` sign before `!pip` in the code cell below.



__Notice :__ Suivez les instructions données dans le premier atelier de ce cours pour créer une instance de service de base de données de Db2 sur le cloud.



## Tâche 1 : Importer la bibliothèque Python 'ibm_db'

The `ibm_db` [API](https://pypi.python.org/pypi/ibm_db/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork22-2022-01-01) fournit une variété de fonctions Python utiles pour accéder aux données d’une base de données de serveur de données IBM® et les manipuler, y compris des fonctions de connexion à une base de données, de préparation et d’émission d’instructions SQL, d’extraction de lignes à partir de jeux de résultats, d’appel de procédures stockées, de validation et d’annulation de transactions, de gestion des erreurs et de récupération de métadonnées.



Nous importons la bibliothèque ibm_db dans notre application Python

Les modules requis suivants sont préinstallés dans l’environnement Skills Network Labs. Toutefois, si vous exécutez ces commandes de bloc-notes dans un autre environnement Jupyter (par exemple, Watson Studio ou Ananconda), vous devrez peut-être installer ces bibliothèques en supprimant le signe # avant !pip dans la cellule de code ci-dessous.

In [ ]:
# These libraries are pre-installed in SN Labs. If running in another environment please uncomment lines below to install them:
# !pip install --force-reinstall ibm_db==3.1.0 ibm_db_sa==0.3.3
# Ensure we don't load_ext with sqlalchemy>=1.4 (incompadible)
# !pip uninstall sqlalchemy==1.4 -y && pip install sqlalchemy==1.3.24
# !pip install ipython-sql

In [1]:
import ibm_db

When the command above completes, the `ibm_db` library is loaded in your notebook. 


## Task 2: Identify the database connection credentials

Connecting to dashDB or DB2 database requires the following information:
* Driver Name
* Database name 
* Host DNS name or IP address 
* Host port
* Connection protocol
* User ID
* User Password



__Notice:__ To obtain credentials please refer to the instructions given in the first Lab of this course

Now enter your database credentials below

Replace the placeholder values in angular brackets <> below with your actual database credentials 

e.g. replace "database" with "BLUDB"



__Notice :__ Pour obtenir des informations d’identification, veuillez vous référer aux instructions données dans le premier laboratoire de ce cours

Entrez maintenant les informations d’identification de votre base de données ci-dessous

Remplacez les valeurs d’espace réservé entre crochets angulaires <> ci-dessous par vos informations d’identification de base de données réelles.


ce lien conduit à mes informations d'identification(j'ai fais les captures d'écran j'ai mis dans mon dossier SQL) qui m'ont été attribué sur le cloud IBM DB2. ME permettant de connecter sur le cloud en utiisant python 

https://author-ide.skills.network/render?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJtZF9pbnN0cnVjdGlvbnNfdXJsIjoiaHR0cHM6Ly9jZi1jb3Vyc2VzLWRhdGEuczMudXMuY2xvdWQtb2JqZWN0LXN0b3JhZ2UuYXBwZG9tYWluLmNsb3VkL0lCTURldmVsb3BlclNraWxsc05ldHdvcmstREIwMjAxRU4tU2tpbGxzTmV0d29yay9sYWJzL01vZHVsZSUyMDUvTEFCLTB2Nl9DcmVhdGVfRGF0YWJhc2VfQ3JlZGVudGlhbHMubWQiLCJ0b29sX3R5cGUiOiJpbnN0cnVjdGlvbmFsLWxhYiIsImFkbWluIjpmYWxzZSwiaWF0IjoxNzAwNjcxNTA5fQ.OenscAhHlBp9pa5GJ7STVJPd_1DzLrSie2apJSqyr9Q

par exemple, remplacez « database » par « BLUDB »


In [ ]:
#Replace the placeholder values with the actuals for your Db2 Service Credentials
dsn_driver = "{IBM DB2 ODBC DRIVER}"
dsn_database = "database"            # e.g. "BLUDB"
dsn_hostname = "hostname"            # e.g.: "dashdb-txn-sbox-yp-dal09-04.services.dal.bluemix.net"
dsn_port = "port"                    # e.g. "50000" 
dsn_protocol = "protocol"            # i.e. "TCPIP"
dsn_uid = "username"                 # e.g. "abc12345"
dsn_pwd = "password"                 # e.g. "7dBZ3wWt9XN6$o0J"
dsn_security = "SSL"              #i.e. "SSL"

In [2]:
dsn_hostname = "2f3279a5-73d1-4859-88f0-a6c3e6b4b907.c3n41cmd0nqnrk39u98g.databases.appdomain.cloud"

dsn_uid = "zyj48746"

dsn_pwd = "n97cCLsRaGSqeB1C"

dsn_driver = "{IBM DB2 ODBC DRIVER}"

dsn_database = "bludb" 

dsn_port = "30756"

dsn_protocol = "TCPIP" 

dsn_security = "SSL" 


## Task 3: Create the database connection

Ibm_db API uses the IBM Data Server Driver for ODBC and CLI APIs to connect to IBM DB2 and Informix.


Create the database connection



## Tâche 3 : Créer la connexion à la base de données

Ibm_db’API utilise le pilote IBM Data Server pour les API ODBC et CLI pour se connecter à IBM DB2 et Informix.

Créer la connexion à la base de données

In [3]:
#Create database connection
#DO NOT MODIFY THIS CELL. Just RUN it with Shift + Enter
dsn = (
    "DRIVER={0};"
    "DATABASE={1};"
    "HOSTNAME={2};"
    "PORT={3};"
    "PROTOCOL={4};"
    "UID={5};"
    "PWD={6};"
    "SECURITY={7};").format(dsn_driver, dsn_database, dsn_hostname, dsn_port, dsn_protocol, dsn_uid, dsn_pwd,dsn_security)

try:
    conn = ibm_db.connect(dsn, "", "")
    print ("Connected to database: ", dsn_database, "as user: ", dsn_uid, "on host: ", dsn_hostname)

except:
    print ("Unable to connect: ", ibm_db.conn_errormsg() )


Connected to database:  bludb as user:  zyj48746 on host:  2f3279a5-73d1-4859-88f0-a6c3e6b4b907.c3n41cmd0nqnrk39u98g.databases.appdomain.cloud


## Task 4: Create a table in the database

In this step we will create a table in the database with following details:

<img src="https://ibm.box.com/shared/static/ztd2cn4xkdoj5erlk4hhng39kbp63s1h.jpg" align="center">


In [5]:
#Lets first drop the table INSTRUCTOR in case it exists from a previous attempt
# supprimez d’abord la table INSTRUCTOR au cas où elle existerait à partir d’une tentative précédente

dropQuery = "drop table INSTRUCTOR"

#Now execute the drop statment


dropStmt = ibm_db.exec_immediate(conn, dropQuery)

Exception: [IBM][CLI Driver][DB2/LINUXX8664] SQL0204N  "ZYJ48746.INSTRUCTOR" is an undefined name.  SQLSTATE=42704 SQLCODE=-204

## Dont worry if you get this error:
If you see an exception/error similar to the following, indicating that INSTRUCTOR is an undefined name, that's okay. It just implies that the INSTRUCTOR table does not exist in the table - which would be the case if you had not created it previously.

Exception: [IBM][CLI Driver][DB2/LINUXX8664] SQL0204N  "ABC12345.INSTRUCTOR" is an undefined name.  SQLSTATE=42704 SQLCODE=-204


## Ne vous inquiétez pas si vous obtenez cette erreur :

Si vous voyez une exception/erreur similaire à la suivante, indiquant que INSTRUCTOR est un nom indéfini, ce n’est pas grave. Cela implique simplement que la table INSTRUCTOR n’existe pas dans la table - ce qui serait le cas si vous ne l’aviez pas créée auparavant.

Exception: [IBM][CLI Driver][DB2/LINUXX8664] SQL0204N  "ABC12345.INSTRUCTOR" is an undefined name.  SQLSTATE=42704 SQLCODE=-204


In [8]:
#Construct the Create Table DDL statement - replace the ... with rest of the statement
#Construct l’instruction Create Table DDL - remplacez le ... avec le reste de l’instruction
createQuery = "create table INSTRUCTOR(ID INTEGER PRIMARY KEY NOT NULL, FNAME VARCHAR(20), LNAME VARCHAR(20), CITY VARCHAR(20), CCODE CHAR(2))"

#Now fill in the name of the method and execute the statement
#Now renseignez le nom de la méthode et exécutez l’instruction

createStmt = ibm_db.replace_with_name_of_execution_method(conn, createQuery)

AttributeError: module 'ibm_db' has no attribute 'replace_with_name_of_execution_method'

In [9]:
#Construct the Create Table DDL statement - replace the ... with rest of the statement
#Construct l’instruction Create Table DDL - remplacez le ... avec le reste de l’instruction
createQuery = "create table INSTRUCTOR(ID INTEGER PRIMARY KEY NOT NULL, FNAME VARCHAR(20), LNAME VARCHAR(20), CITY VARCHAR(20), CCODE CHAR(2))"

#Now fill in the name of the method and execute the statement
#Now renseignez le nom de la méthode et exécutez l’instruction

createStmt = ibm_db.exec_immediate(conn,createQuery)


<details><summary>Click here for the solution</summary>

```python
createQuery = "create table INSTRUCTOR(ID INTEGER PRIMARY KEY NOT NULL, FNAME VARCHAR(20), LNAME VARCHAR(20), CITY VARCHAR(20), CCODE CHAR(2))"

createStmt = ibm_db.exec_immediate(conn,createQuery)
```

</details>


## Task 5: Insert data into the table

In this step we will insert some rows of data into the table. 

The INSTRUCTOR table we created in the previous step contains 3 rows of data:

<img src="https://ibm.box.com/shared/static/j5yjassxefrjknivfpekj7698dqe4d8i.jpg" align="center">

We will start by inserting just the first row of data, i.e. for instructor Rav Ahuja 


## Tâche 5 : Insérer des données dans le tableau

Dans cette étape, nous allons insérer quelques lignes de données dans le tableau. 

La table INSTRUCTOR que nous avons créée à l’étape précédente contient 3 lignes de données :

<img src="https://ibm.box.com/shared/static/j5yjassxefrjknivfpekj7698dqe4d8i.jpg" align="center">


Nous commencerons par insérer uniquement la première ligne de données, c’est-à-dire pour l’instructeur Rav Ahuja

In [10]:
#Construct the query - replace ... with the insert statement
insertQuery = " insert into INSTRUCTOR values (1, 'Rav', 'Ahuja', 'TORONTO', 'CA')"

#execute the insert statement
insertStmt = ibm_db.exec_immediate(conn, insertQuery)

<details><summary>Click here for the solution</summary>

```python
insertQuery = "insert into INSTRUCTOR values (1, 'Rav', 'Ahuja', 'TORONTO', 'CA')"

insertStmt = ibm_db.exec_immediate(conn, insertQuery)
```

</details>


Now use a single query to insert the remaining two rows of data

Utilisez maintenant une seule requête pour insérer les deux lignes de données restantes


In [12]:
#replace ... with the insert statement that inerts the remaining two rows of data
insertQuery2 = "insert into INSTRUCTOR values (2, 'Raul', 'Chong', 'Markham', 'CA'),  (3, 'Hima', 'Vasudevan', 'Chicago', 'US')"

#execute the statement
insertStmt2 = ibm_db.exec_immediate(conn, insertQuery2)

<details><summary>Click here for the solution</summary>

```python
insertQuery2 = "insert into INSTRUCTOR values (2, 'Raul', 'Chong', 'Markham', 'CA'), (3, 'Hima', 'Vasudevan', 'Chicago', 'US')"

insertStmt2 = ibm_db.exec_immediate(conn, insertQuery2)

```

</details>


## Task 6: Query data in the table

In this step we will retrieve data we inserted into the INSTRUCTOR table. 


## Tâche 6 : Interroger les données dans la table

Dans cette étape, nous allons récupérer les données que nous avons insérées dans la table INSTRUCTOR. 

In [14]:
#Construct the query that retrieves all rows from the INSTRUCTOR table
selectQuery = "select * from INSTRUCTOR"

#Execute the statement
selectStmt = ibm_db.exec_immediate(conn, selectQuery)

#Fetch the Dictionary (for the first row only) - replace ... with your code
#Fetch le dictionnaire (pour la première ligne uniquement) - remplacez ... avec votre code
ibm_db.fetch_both(selectStmt)

{'ID': 1,
 0: 1,
 'FNAME': 'Rav',
 1: 'Rav',
 'LNAME': 'Ahuja',
 2: 'Ahuja',
 'CITY': 'TORONTO',
 3: 'TORONTO',
 'CCODE': 'CA',
 4: 'CA'}

# Construct the query that retrieves all rows from the INSTRUCTOR table selectQuery = "select * from INSTRUCTOR" #Execute the statement selectStmt = ibm_db.exec_immediate(conn, selectQuery) #Fetch the Dictionary (for the first row only) - replace ... with your code ibm_db.fetch_both(selectStmt)
Le segment de code est écrit en Python et est utilisé pour interagir avec une base de données à l'aide du pilote de base de données IBM Db2 pour Python (ibm_db).

Construction de la requête :

selectQuery = "select * from INSTRUCTOR":

Cette ligne construit une chaîne de requête SQL pour récupérer toutes les lignes de la table INSTRUCTOR. Le * dans l'instruction SELECT signifie que toutes les colonnes seront récupérées.


Exécution de la déclaration :

selectStmt = ibm_db.exec_immediate(conn, selectQuery) :

Cette ligne exécute la requête SQL spécifiée dans selectQuery. Il utilise la fonction ibm_db.exec_immediate() pour exécuter la requête immédiatement sans préparer ni lier de paramètres. Le paramètre conn est l'objet de connexion représentant la connexion à la base de données.

Récupération du résultat :

ibm_db.fetch_both(selectStmt) :

Cette ligne récupère la ligne suivante d'un jeu de résultats produit par une instruction SQL précédemment exécutée. Ici, il récupère la première ligne du jeu de résultats renvoyé par selectQuery. La fonction ibm_db.fetch_both() renvoie un dictionnaire contenant les valeurs des colonnes de la ligne récupérée. Le dictionnaire contient des indices de tableau numériques et associatifs, permettant d'accéder aux valeurs de colonne en utilisant à la fois les noms de colonnes et les indices numériques.



En résumé, ce segment de code construit une requête SQL pour récupérer toutes les lignes de la table INSTRUCTOR, exécute la requête, puis récupère la première ligne du jeu de résultats sous forme de dictionnaire.

Explication du résultat :

Cette sortie représente une seule ligne du jeu de résultats renvoyé par la requête de base de données. Il est structuré comme un dictionnaire où les clés correspondent aux noms de colonnes et les valeurs correspondent aux données récupérées pour chaque colonne de la ligne.




- 'ID' (ou 0) : La valeur 1 représente la valeur de la colonne ID dans la table de la base de données. La touche 0 est également fournie sous forme d'index numérique pour accéder à la même valeur.

- 'FNAME' (ou 1) : La valeur 'Rav' représente la valeur de la colonne FNAME dans la table de la base de données. La touche 1 est également fournie sous forme d'index numérique pour accéder à la même valeur.

- 'LNAME' (ou 2) : La valeur 'Ahuja' représente la valeur de la colonne LNAME dans la table de la base de données. La clé 2 est également fournie sous forme d'index numérique pour accéder à la même valeur.

- 'CITY' (ou 3) : La valeur 'TORONTO' représente la valeur de la colonne CITY dans la table de la base de données. La touche 3 est également fournie sous forme d'index numérique pour accéder à la même valeur.

- CCODE' (ou 4) : La valeur 'CA' représente la valeur de la colonne CCODE dans la table de la base de données. La touche 4 est également fournie sous forme d'index numérique pour accéder à la même valeur.



En résumé, ce dictionnaire offre un moyen flexible d'accéder aux données extraites de la requête de base de données, permettant d'accéder aux valeurs de colonne en utilisant à la fois les noms de colonne et les index numériques. Cela peut être particulièrement utile lorsque vous travaillez avec des requêtes SQL générées dynamiquement ou lorsque les noms de colonnes ne sont pas connus à l'avance.

J'espère que ça aide.

Bon apprentissage!


<details><summary>Click here for the solution</summary>

```python
#Construct the query that retrieves all rows from the INSTRUCTOR table
selectQuery = "select * from INSTRUCTOR"

#Execute the statement
selectStmt = ibm_db.exec_immediate(conn, selectQuery)

#Fetch the Dictionary (for the first row only)
ibm_db.fetch_both(selectStmt)

```

</details>


In [15]:
#Fetch the rest of the rows and print the ID and FNAME for those rows
#Fetch le reste des lignes et imprimez l’ID et le FNAME de ces lignes

while ibm_db.fetch_row(selectStmt) != False:
   print (" ID:",  ibm_db.result(selectStmt, 0), " FNAME:",  ibm_db.result(selectStmt, "FNAME"))

 ID: 2  FNAME: Raul
 ID: 3  FNAME: Hima


Salut @MAHI,

Merci de votre publication.

Nous sommes heureux de vous assister.

Nous pensons que vous avez besoin d'une explication pour le bloc de code fourni. Laisse-moi expliquer:

#Fetch the rest of the rows and print the ID and FNAME for those rows while - ---    -- -ibm_db.fetch_row(selectStmt) != False: print (" ID:", ibm_db.result(selectStmt, 0), " FNAME:", - ibm_db.result(selectStmt, "FNAME"))
Ce segment de code est utilisé pour récupérer et imprimer les lignes restantes du jeu de résultats renvoyé par une requête de base de données. Voici une explication de chaque partie du code :

Boucle While :

Cette ligne initie une boucle while qui continue tant qu'il y a plus de lignes à extraire du jeu de résultats. La fonction ibm_db.fetch_row(selectStmt) récupère la ligne suivante du jeu de résultats et la boucle continue jusqu'à ce que cette fonction renvoie False, indiquant qu'il n'y a plus de lignes à extraire.

Imprimer la déclaration :

Dans la boucle, cette ligne imprime les valeurs des colonnes ID et FNAME pour chaque ligne récupérée.

- ibm_db.result(selectStmt, 0) récupère la valeur de la colonne ID pour la ligne actuelle. Le paramètre 0 représente l'index de la colonne dans le jeu de résultats.

- ibm_db.result(selectStmt, "FNAME") récupère la valeur de la colonne FNAME pour la ligne actuelle. Le paramètre "FNAME" représente le nom de la colonne dans le jeu de résultats.

En résumé, ce segment de code parcourt l'ensemble de résultats récupéré par le curseur selectStmt et imprime les valeurs ID et FNAME pour chaque ligne. Il s'agit d'un modèle courant utilisé pour traiter et afficher les résultats d'une requête ligne par ligne.

J'espère que ça aide.

Bon apprentissage!

<details><summary>Click here for the solution</summary>

```python
#Fetch the rest of the rows and print the ID and FNAME for those rows
while ibm_db.fetch_row(selectStmt) != False:
    print (" ID:",  ibm_db.result(selectStmt, 0), " FNAME:",  ibm_db.result(selectStmt, "FNAME"))

```

</details>


Bonus: now write and execute an update statement that changes the Rav's CITY to MOOSETOWN 

Bonus : écrivez et exécutez maintenant une déclaration de mise à jour qui change la VILLE du Rav en MOOSETOWN


In [16]:
#Enter your code below
updateQuery = "update INSTRUCTOR set CITY='MOOSETOWN' where FNAME='Rav'"
updateStmt = ibm_db.exec_immediate(conn, updateQuery)


<details><summary>Click here for the solution</summary>

```python
updateQuery = "update INSTRUCTOR set CITY='MOOSETOWN' where FNAME='Rav'"
updateStmt = ibm_db.exec_immediate(conn, updateQuery)

```

</details>


## Task 7: Retrieve data into Pandas 

In this step we will retrieve the contents of the INSTRUCTOR table into a Pandas dataframe


In [17]:
import pandas
import ibm_db_dbi

In [18]:
#connection for pandas
pconn = ibm_db_dbi.Connection(conn)

In [19]:
#query statement to retrieve all rows in INSTRUCTOR table
selectQuery = "select * from INSTRUCTOR"

#retrieve the query results into a pandas dataframe
pdf = pandas.read_sql(selectQuery, pconn)

#print just the LNAME for first row in the pandas data frame
pdf.LNAME[0]

'Ahuja'

In [20]:
#print the entire data frame
pdf

ID FNAME      LNAME       CITY CCODE
0   1   Rav      Ahuja  MOOSETOWN    CA
1   2  Raul      Chong    Markham    CA
2   3  Hima  Vasudevan    Chicago    US

Once the data is in a Pandas dataframe, you can do the typical pandas operations on it. 

For example you can use the shape method to see how many rows and columns are in the dataframe

Une fois que les données sont dans un dataframe Pandas, vous pouvez effectuer les opérations typiques de pandas sur celui-ci. 

Par exemple, vous pouvez utiliser la méthode shape pour voir combien de lignes et de colonnes se trouvent dans le dataframe



In [21]:
pdf.shape

(3, 5)

## Task 8: Close the Connection
We free all resources by closing the connection. Remember that it is always important to close connections so that we can avoid unused connections taking up resources.


In [22]:
ibm_db.close(conn)

True

## Summary

In this tutorial you established a connection to a database instance of DB2 Warehouse on Cloud from a Python notebook using ibm_db API. Then created a table and insert a few rows of data into it. Then queried the data. You also retrieved the data into a pandas dataframe.


Dans ce didacticiel, vous avez établi une connexion à une instance de base de données de DB2 Warehouse on Cloud à partir d’un notebook Python à l’aide de ibm_db API. Ensuite, vous avez créé une table et insérez-y quelques lignes de données. Ensuite, interrogez les données. Vous avez également récupéré les données dans une trame de données pandas.

## Author

<a href="https://www.linkedin.com/in/ravahuja/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork22-2022-01-01" target="_blank">Rav Ahuja</a>



## Change Log


|  Date (YYYY-MM-DD) |  Version | Changed By  |  Change Description |
|---|---|---|---|
| 2021-11-17  | 2.2  | Lakshmi  | Updated library  |
| 2021-07-09  | 2.1  | Malika   | Updated connection string   |
| 2020-08-28  | 2.0  | Lavanya  |  Moved lab to course repo in GitLab |



<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
